projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-plugins-modify-endpoint.php

```diff
@@ -346,6 +346,12 @@ protected function update() {
 				continue;
 			}
 
+			// Establish per plugin lock.
+			$plugin_slug = Jetpack_Autoupdate::get_plugin_slug( $plugin );
+			if ( ! WP_Upgrader::create_lock( 'jetpack_' . $plugin_slug ) ) {
```

Comment:
If this PR does not remove the 'auto_updater' lock creation above, we will continue running into conflicts when trying to update multiple plugins.

That said, maybe it would be good to try avoiding conflicts with core plugin update attempts by:
1. Stop from doing things when a core update is already in progress: Stop creating the 'auto_updater' lock but still refuse to update if the lock exists and has not timed out
2. Stop core from updating while a Jetpack plugin update is already in progress: Filter the 'update_plugins' transient used [here](https://github.com/WordPress/wordpress-develop/blob/15b43fb36098dcd4785520b60ec082c9ff87de1c/src/wp-admin/includes/class-wp-automatic-updater.php#L465), removing all plugins for which Jetpack has a plugin-specific update lock

Reply:
Hmm yeah I was thinking we would leave the auto_updater in place as it would allow us to short circuit core without any further action but it won't solve the issue with the Calypso calls where it makes multiple calls to update plugins. Your recommendation makes sense with the addition that 

3. we still lock 'auto_updater' if this is an Jetpack autoupdate request. Constants::get_constant( 'JETPACK_PLUGIN_AUTOUPDATE' )

Reply:
If I think on it further though this getscomplicated as we would need a permanent filter based on if a lock exists so we would have to check for 'jetpack-*.lock' options for every plugin in that update_plugins value. In addition the code would be brittle to changes in WP.com as no "lock_exists" function exists in WP_Upgrader.

I don't think we want to go down that road. Instead I think we do a hybrid

1) AutoUpdate request Constants::get_constant( 'JETPACK_PLUGIN_AUTOUPDATE' ) -> do auto_updater lock , no individual plugin locks needed.
2) not autoUpdate request -> check the auto_updater lock (create/release) do individual plugin lock.

I think this is one of those cases where it removes most conflicts and it doesn't add undo checks per the update_plugins site transient filter.

Reply:
I may need to let what you wrote sink in a bit more, but currently, it seems like my first two suggestions should be sufficient for avoiding conflicts with core plugin auto-updates.

As for avoiding conflicts when Jetpack makes multiple update requests, I think per-plugin locking should take care of all cases. We already have locking on the WordPress.com side to prevent multiple, simultaneous bulk update attempts. If one of those attempts overlaps with individual update attempts from Calypso, per-plugin locking should still prevent simultaneous updates from breaking things.

> If I think on it further though this getscomplicated as we would need a permanent filter based on if a lock exists so we would have to check for 'jetpack-*.lock' options for every plugin in that update_plugins value. In addition the code would be brittle to changes in WP.com as no "lock_exists" function exists in WP_Upgrader.

The filter could return early if the `auto_updater.lock` is not set with a fresh enough timestamp. Then, if core's global update lock is indeed in place, we would have to check for a lock option for each plugin. This is definitely coupled to the WP_Upgrader lock implementation, but if you are worried about the implementation changing, you could always write your own locking functions that work the same way. Since these locks are internal to Jetpack, maybe that makes sense anyway.

> 3. we still lock 'auto_updater' if this is an Jetpack autoupdate request. Constants::get_constant( 'JETPACK_PLUGIN_AUTOUPDATE' )

Why is this needed?

My mind is a bit foggy at the end of the day, so I apologize if I'm misreading or missing something obvious. Will take another look in the morning.

Reply:
I've gone through this several times and I do believe we want to restrict the behavior to Jetpack. If we truely wanted to lock down WordPress so that if an individual plugin is being updated no other actions can trigger the same update that would need to be a core update. Adding in logic to Plugin_Upgrader so that it does locking and checking that if a plugin is being updated that it returns out with proper messaging.

I think trying to fully lock down plugin updates through Jetpack is out of scope.

What Jetpack should focus on is playing nicely with core and ensuring its update flows/routines do not lead to conflict which is what we get with the current implementation.

1) Core and Jetpack's  auto update routines are now exclusive. By using the same lock they will not clash

2) Jetpack update endpoint locks per plugin. This means if Jetpack's auto update routine is running and it is updating a plugin a 2nd request to update that plugin will not be allowed. Same if an update of a plugin is occurring and the auto-update triggers then the auto-update will skip locked plugins. Other flows like User Activity that trigger multiple requests will succeed as long as no other Jetpack process is updating the plugin.

I think this is the scope that Jetpack should be concerned with. We are ensuring that our flows don't conflict, However if users do other activities outside of Jetpack (Plugin screen, cli, other plugins) we aren't interfering.

Reply:
Thank you for explaining. I think you're right about the scope of this PR, and your changes make good sense to me.

Outside of the scope of this PR (and sometime later), I think Jetpack will still eventually need to adjust core plugin auto-updates so they do not interfere with Jetpack's plugin update rollbacks.